File: cpp/leetcode/216.cpp
class Solution {
public:
    vector<vector<int>> combinationSum3(int k, int n) {
        vector<int> ans;
        vector<vector<int>> res;
        help(1, k, n, ans, res);
        return res;
    }
    
    void help(int start, int k, int n, vector<int> &ans, vector<vector<int>> &res){
        if(k == 0 && n == 0){
            res.push_back(ans);
            return;
        }
        for(int i = start; i <= 9; i++){
            ans.push_back(i);
            help(i + 1, k - 1, n - i, ans, res);
            ans.pop_back();
        }
    }
};

File: python/leetcode/216.py
class Solution(object):
    def combinationSum3(self, k, n):
        if k == 0: return []
        res = []
        self.help(k, n, 0, 0, [], res)
        return res
    
    def help(self, k, n, start, tmp, ans, res):
        if len(ans) > k or tmp > n:
            return 
        if len(ans) == k and tmp == n:
            res.append(ans)
            return
        for i in range(start + 1, 10):
            self.help(k, n, i, tmp + i, ans + [i], res)